WIP Ops 5990 autoscale on jobs

dlobue opened the pull request:
OPS-5990

dlobue committed:
First pass at job autoscaling

OPS-5990

dlobue committed:
Add gh workflow to run tests on PRs

Kami commented on controllers/autoscaling.go:
```diff
@@ -52,18 +52,37 @@ func (r *HorizontalRunnerAutoscalerReconciler) determineDesiredReplicas(rd v1alp
+				jobs, _, err := r.GitHubClient.Actions.ListWorkflowJobs(context.TODO(), user, repoName, run.GetID(), nil)
+				if err != nil {
+					r.Log.Error(err, "Error listing workflow jobs")
+					inProgress++
```
If I understand this correctly, this means that if a workflow has 4 jobs and all 4 are running / in progress, ``inProgress`` will actually be set to 5 - 1 for the whole workflow and 4 for the child jobs?

Or am I misunderstanding it?

dlobue replied:
No, `inProgress` will only be 4. We do not count the workflow itself when we successfully list the wf's jobs. The only time workflows are represented in the `inProgress` variable is when an error is encountered while trying to list the child jobs, or when the wf has no child jobs for some reason.

Kami commented on controllers/autoscaling.go:
```diff
@@ -52,18 +52,37 @@ func (r *HorizontalRunnerAutoscalerReconciler) determineDesiredReplicas(rd v1alp
 				completed++
 			case "in_progress":
-				inProgress++
+				jobs, _, err := r.GitHubClient.Actions.ListWorkflowJobs(context.TODO(), user, repoName, run.GetID(), nil)
```
In general the change looks good to me and should be fine as long as we don't monitor too many repos with many jobs, otherwise we would indeed likely hit the API rate limit fast.

But yeah, I checked the Github API and there is not much more we can do when utilizing the polling approach.

One additional thing / improvement we could do is utilize ``If-Modified-Since`` header (https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#conditional-requests).

This way request would not count against rate limit quota if nothing has changed, but then again I don't think we would hit that scenario all that often in this list workflows and list workflow jobs scenario which is pretty dynamic / changing in nature (although it also kinds depends on the polling interval).

Kami replied:
I had a quick look at the go client - https://github.com/google/go-github/blob/master/github/actions_workflows.go#L65

It looks like we would need to modify the client to support passing custom headers to utilize If-Modified-Since functionality  (which may be a good upstream contribution anyway).

dlobue replied:
Good call on the `If-Modified-Since` header, I'll add that!

I plan to iterate on this and eventually add a webhook and just use those events to scale.

dlobue committed:
Include the number of jobs on queued workflows in autoscale decision

Include the number of jobs on queued workflows in autoscale decision. If
we don't, we'll just end up autoscaling the next time we poll anyway.

OPS-5990

dlobue committed:
Add unit tests for job autoscaler

OPS-5990
